src/machines/onboardingMachine.tsx

```diff
@@ -1,47 +1,99 @@
-  ctx.termsConditionsAccepted && ctx.privacyPolicyAccepted;
-
-const onboardingPersonalDataGuard = (ctx: OnboardingMachineContext, _event: OnboardingMachineEventTypes) => {
-  const {firstName, lastName, emailAddress} = ctx.personalData;
-  return firstName && firstName.length > 0 && lastName && lastName.length > 0 && emailAddress && EMAIL_ADDRESS_VALIDATION_REGEX.test(emailAddress);
-};
-
-const onboardingPinCodeSetGuard = (ctx: OnboardingMachineContext, _event: OnboardingMachineEventTypes) => {
-  const {pinCode} = ctx;
-  return pinCode && pinCode.length === 6;
-};
-
-const onboardingPinCodeVerifyGuard = (ctx: OnboardingMachineContext, event: NextEvent) => {
-  return onboardingPinCodeSetGuard(ctx, event) && ctx.pinCode === event.data;
+const states: OnboardingStatesConfig = {
+  showIntro: {
+    on: {
+      NEXT: OnboardingMachineStateType.showProgress,
+    },
+  },
+  showProgress: {
+    on: {
+      NEXT: [
+        {cond: ({currentStep}) => currentStep === 1, target: OnboardingMachineStateType.enterName},
+        {cond: ({currentStep}) => currentStep === 2, target: OnboardingMachineStateType.enterPinCode},
+      ],
+      PREVIOUS: [
+        {cond: ({currentStep}) => currentStep === 1, target: OnboardingMachineStateType.showIntro},
+        {
+          cond: ({currentStep}) => currentStep === 2,
+          target: OnboardingMachineStateType.enterCountry,
+          actions: assign({currentStep: 1}),
+        },
+      ],
+    },
+  },
+  enterName: {
+    on: {
+      NEXT: OnboardingMachineStateType.enterEmailAddress,
+      PREVIOUS: OnboardingMachineStateType.showProgress,
+      SET_NAME: {actions: assign({name: (_, event) => event.data})},
+    },
+  },
+  enterEmailAddress: {
+    on: {
+      NEXT: OnboardingMachineStateType.enterCountry,
+      PREVIOUS: OnboardingMachineStateType.enterName,
+      SET_EMAIL_ADDRESS: {actions: assign({emailAddress: (_, event) => event.data})},
+    },
+  },
+  enterCountry: {
+    on: {
+      NEXT: {
+        target: OnboardingMachineStateType.showProgress,
+        actions: assign({currentStep: 2}),
+      },
+      PREVIOUS: OnboardingMachineStateType.enterEmailAddress,
+      SET_COUNTRY: {actions: assign({country: (_, event) => event.data})},
+    },
+  },
+  enterPinCode: {
+    on: {
+      NEXT: OnboardingMachineStateType.verifyPinCode,
+      PREVIOUS: OnboardingMachineStateType.showProgress,
+      SET_PIN_CODE: {actions: assign({pinCode: (_, event) => event.data})},
+    },
+  },
+  verifyPinCode: {
+    on: {
+      NEXT: OnboardingMachineStateType.enableBiometrics,
+      PREVIOUS: OnboardingMachineStateType.enterPinCode,
+    },
+  },
+  enableBiometrics: {
+    on: {
+      NEXT: OnboardingMachineStateType.acceptTermsAndPrivacy,
+      PREVIOUS: OnboardingMachineStateType.verifyPinCode,
+    },
+  },
+  acceptTermsAndPrivacy: {
+    on: {
+      READ_TERMS: OnboardingMachineStateType.readTermsAndPrivacy,
```

Comment:
i guess this is for showing the terms itself? should we not have one for the privacy as well?

Reply:
You are right. I didnt notice the two separate screens for them, and assumed they were combined [here](https://www.figma.com/design/6VOB3zOLuRJvGrgGme9Ips/MWALL---Design-file?node-id=761-26829&t=7yyOR6kfIOfNrnHl-0). I will add the extra states

Reply:
Fixed. Query params pass down the state, so a user can see the relevant document when visiting the "read" page